segtypes/common/codesubsegment.py

```diff
@@ -80,545 +41,163 @@ def needs_symbols(self) -> bool:
+        func_sym = self.parent.create_symbol(func_spim.vram, type="func", define=True)
+        func_sym.given_name = func_spim.name
 
-        end_func = False
-        start_new_func = True
-        labels = []
+        # Gather symbols found by spimdisasm and create those symbols in splat's side
+        for referencedVram in func_spim.referencedVRams:
+            context_sym = symbols.spim_context.getAnySymbol(referencedVram)
+            if context_sym is not None:
+                if context_sym.type == spimdisasm.common.SymbolSpecialType.branchlabel:
+                    continue
+                sym_type = None
+                if context_sym.type == spimdisasm.common.SymbolSpecialType.jumptable:
+                    sym_type = "jtbl"
+                    self.parent.jumptables[referencedVram] = (
+                        func_spim.vram,
+                        func_spim.vramEnd,
+                    )
+                sym = self.parent.create_symbol(
+                    referencedVram, type=sym_type, reference=True
+                )
+                sym.given_name = context_sym.getName()
 
-        big_endian = options.get_endianess() == "big"
+        for label_offset in func_spim.localLabels:
+            label_vram = func_spim.getVramOffset(label_offset)
+            label_sym = self.parent.get_symbol(
+                label_vram, type="label", reference=True, local_only=True
+            )
 
-        # Collect labels
-        for insn in insns:
-            if self.is_branch_insn(insn.mnemonic):
-                op_str_split = insn.op_str.split(" ")
-                branch_target = op_str_split[-1]
-                branch_addr = int(branch_target, 0)
-                labels.append((insn.address, branch_addr))
+            if label_sym is not None:
+                context_sym = symbols.spim_context.getAnySymbol(label_vram)
+                if context_sym is not None:
+                    context_sym.name = label_sym.name
+            else:
+                self.parent.labels_to_add.add(label_vram)
 
         # Main loop
-        for i, insn in enumerate(insns):
-            mnemonic = insn.mnemonic
-            op_str = insn.op_str
-
-            # If this is non-zero, disasm size insns
-            hard_size = 0
-
-            if start_new_func:
-                func: Symbol = self.parent.create_symbol(insn.address, type="func")
-                start_new_func = False
+        for i, insn in enumerate(func_spim.instructions):
+            instr_offset = i * 4
+            insn_address = func_sym.vram_start + instr_offset
 
-            if func.size > 4:
-                hard_size = func.size / 4
-
-            if options.get_compiler() == SN64:
-                op_str = self.replace_reg_names(op_str)
+            if insn == spimdisasm.mips.instructions.InstructionId.JR:
+                # Record potential jtbl jumps
+                rs = insn.getRegisterName(insn.rs)
+                if rs not in ["$ra", "$31"]:
```

Comment:
```suggestion
                if rs not in {"$ra", "$31"}:
```
(it it unexpectedly faster usually to use a set)

(also this is more a spimdisasm comment but I think there should be an enum for the integers instead of going to the string names)

Reply:
oh yeah i had forgotten about this bit. It is the minimal adaptation from what splat already had:
https://github.com/ethteck/splat/blob/da54efcfc565c629069e5d61b600a4ae3eae0ffe/segtypes/common/codesubsegment.py#L242-L245

I considered enums at some point, but you can't just compare a python enum to a raw number (for example `insn.rs` and family). Supporting multiple ABIs can make this a bit more complex too.

Reply:
Could you maybe do something like:
`insn.rs not in { MipsO32Regs.RA, MipsN32Regs.RA }`
where `MipsO32Regs` and `MipsN32Regs` are `IntEnum` which allows comparison with integers. Ideally the python recompiler would recognize that both are constants valued at `31` and de-duplicate the set elements, but I'm not sure if that would actually happen...
Just a thought

Reply:
CPython is not smart enough to treat `MipsO32Regs.RA` as a constant: it would require invalidation logic for the case where someone overwrites either `MipsO32Regs` or `MipsO32Regs.RA` which CPython lacks. (Pypy on the other hand is able to do it)

(I also find `... in { MipsO32Regs.RA, MipsN32Regs.RA }` a bit scary personally: it would have false positives if o32 and n32 register numbers for $ra weren't the same)

Reply:
This is the kind of problem why I didn't use enums for registers.

I think I'll go with a raw number and a comment saying it is register `$ra` for now until I figure out a better way to handle registers

Reply:
Actually... I just realized `jtbl_jumps` is not used anywhere else besides here.
Seems like it was used by the old code I removed... So I'll delete it too :sweat_smile: